### ISSUE TITLE
WebGL Renderer is showing black tiles

### ISSUE THREAD

**Issue by thec0keman** [openseadragon/openseadragon#2470]
We are experimenting with the new WebGL renderer on the master branch.

We have OSD rendering our tile sources correctly when `drawer: 'canvas'`.  However, as soon as we switch to `webgl` the tiles all appear as completely black.

This is using a custom tile source (we have a custom implementation for `getTileUrl`), and the tiles are correctly fetching in the browser.  It looks like the image data is correctly making it to `_onTileLoad`, so it doesn't seem likely this has something to do with our implementation of `getTileUrl`.  With debug mode on we can further confirm that the correct metadata about tiles / levels are there, they just aren't rendering correctly.

This is a basic demonstration of what we're doing:
```
const viewer = OpenSeadragon({
  debugMode: true,
  loadTilesWithAjax: false,
  crossOriginPolicy: false,
  drawer: 'webgl',
  id: 'contentDiv',
  tileSources: [
    {
      tileSource: {
        getTileUrl,
        height: 65536,
        width: 131072,
        tileSize: 256,
        tileHeight: 256,
        tileWidth: 256,
        minLevel: 0,
        maxLevel: 8,
      },
    },
  ],
});
```
![Screenshot 2024-02-09 at 10 41 44 AM](https://github.com/openseadragon/openseadragon/assets/4037898/37d2bb96-3e29-4a94-8da7-3cbeaff5412a)

Is there some additional configuration we need to add to get the webgl renderer working with a custom tile source?

**Comment by Aiosa** [comment]
Can you provide logs from the console or a reproducible example (e.g. your tilesource + possibly accees to the data if required)? Hard to tell just from this. WebGL1 has some issues with constrains on the texture data, I remember being once forced to convert my textures to rectangles to correctly upload them, it might be a problem if your data is not perfectly rectangular. So hard to say from such little information. WebGL2 support + modularization are planned.

**Comment by pearcetm** [comment]
I've seen this with cross-origin data. WebGL doesn't let you upload data to the graphics card from a tainted canvas/cross-origin image.

Try `crossOrigin: 'Anonymous'` in your config and see if that helps. If the server doesn't allow that, I think you will need to stick to `drawer: 'canvas'`.

**Comment by pearcetm** [comment]
I realized this is likely to be a fairly common occurrence, and that there is a fairly straightforward to handle it with our current architecture.

I created a PR that detects when the `WebGLDrawer` cannot draw a `TiledImage` because of tainted data (cross-origin without CORS enabled), and falls back to using the `CanvasDrawer` specifically for those images.

Of course, you'll lose all the benefits of the `WebGLDrawer` like absence of tiling artifacts, filters (in the future), etc. But at least the viewer won't just fail to render.

**Comment by iangilman** [comment]
BTW, @thec0keman thank you for testing the new code!

**Comment by thec0keman** [comment]
Fix for CORS worked great.  Thanks!

**Comment by Aiosa** [comment]
What about this? https://blog.chromium.org/2011/07/using-cross-domain-images-in-webgl-and.html?m=1

**Comment by pearcetm** [comment]
@Aiosa I believe what the blog post you linked is saying is that the image needs to be fetched with CORS enabled in order to be not tainted, and therefore be usable by WebGL. This is done on an `<img>` tag by setting `crossOrigin = 'Anonymous'`, which is what we do in OSD with `crossOriginPolicy: 'Anonymous'`. The server also has to be configured to allow this.

This is what I suggested above, and it sounds like it worked for @thec0keman. Cross-origin data is just fine as long as CORS is used, and is drawn by WebGLDrawer without a problem. The internal canvas drawer accounts for the other scenario - when CORS is disallowed by the server (or by a dev who doesn't use the `crossOriginPolicy` setting for whatever reason). In this case, the `CanvasDrawer` can still display the data (but the canvas itself will be tainted and won't allow access to the pixel data).

**Comment by pearcetm** [comment]
> we can either set the attribute (having IMG) or convert to data URL and back (having canvas)

@Aiosa if the data is fetched from a cross-origin site without CORS, you can't convert to data URL because the canvas will have already been tainted. You also can't change the `crossOrigin` attribute on the `<img>` once it is loaded to get around this (or at least it won't have any effect).

**Comment by iangilman** [comment]
I believe this is now fixed by #2472.
